src/PoolManager.sol

```diff
@@ -197,47 +198,66 @@ contract PoolManager is IPoolManager, ProtocolFees, NoDelegateCall, ERC6909Claim
             })
         );
 
-        _accountPoolBalanceDelta(key, delta + feeDelta);
+        callerDelta = principalDelta + feesAccrued;
 
         emit ModifyLiquidity(id, msg.sender, params.tickLower, params.tickUpper, params.liquidityDelta);
 
-        key.hooks.afterModifyLiquidity(key, params, delta, hookData);
+        // if the hook doesnt have the flag to be able to return deltas, hookDelta will always be 0.
+        BalanceDelta hookDelta;
+        (callerDelta, hookDelta) = key.hooks.afterModifyLiquidity(key, params, callerDelta, hookData);
+
+        if (hookDelta != BalanceDeltaLibrary.ZERO_DELTA) _accountPoolBalanceDelta(key, hookDelta, address(key.hooks));
+
+        _accountPoolBalanceDelta(key, callerDelta, msg.sender);
     }
 
     /// @inheritdoc IPoolManager
     function swap(PoolKey memory key, IPoolManager.SwapParams memory params, bytes calldata hookData)
         external
         override
         onlyWhenUnlocked
-        returns (BalanceDelta)
+        returns (BalanceDelta swapDelta)
     {
+        if (params.amountSpecified == 0) revert SwapAmountCannotBeZero();
+
         PoolId id = key.toId();
         _checkPoolInitialized(id);
 
-        key.hooks.beforeSwap(key, params, hookData);
+        (int256 amountToSwap, int128 hookDeltaSpecified) = key.hooks.beforeSwap(key, params, hookData);
 
-        (BalanceDelta delta, uint256 feeForProtocol, uint24 swapFee, Pool.SwapState memory state) = pools[id].swap(
+        // execute swap, account protocol fees, and emit swap event
+        swapDelta = _swap(
+            id,
             Pool.SwapParams({
                 tickSpacing: key.tickSpacing,
                 zeroForOne: params.zeroForOne,
-                amountSpecified: params.amountSpecified,
+                amountSpecified: amountToSwap,
                 sqrtPriceLimitX96: params.sqrtPriceLimitX96
-            })
+            }),
+            params.zeroForOne ? key.currency0 : key.currency1
         );
 
-        _accountPoolBalanceDelta(key, delta);
+        BalanceDelta hookDelta;
+        (swapDelta, hookDelta) = key.hooks.afterSwap(key, params, swapDelta, hookData, hookDeltaSpecified);
+
+        // if the hook doesnt have the flag to be able to return deltas, hookDelta will always be 0
+        if (hookDelta != BalanceDeltaLibrary.ZERO_DELTA) _accountPoolBalanceDelta(key, hookDelta, address(key.hooks));
+
+        _accountPoolBalanceDelta(key, swapDelta, msg.sender);
+    }
+
+    // Internal swap function to execute a swap, account protocol fees, and emit the swap event
+    function _swap(PoolId id, Pool.SwapParams memory params, Currency inputCurrency) internal returns (BalanceDelta) {
```

Comment:
ie here we use inputCurrency (which honestly I'm leaning to like more than specifiedCurrency?) but up to you